library/src/scala/runtime/Invoker.scala

```diff
@@ -34,40 +27,24 @@ object Invoker {
     * @param dataDir
     *   the directory where the measurement data is held
     */
-  def invoked(id: Int, dataDir: String): Unit = {
-    // [sam] we can do this simple check to save writing out to a file.
-    // This won't work across JVMs but since there's no harm in writing out the same id multiple
-    // times since for coverage we only care about 1 or more, (it just slows things down to
-    // do it more than once), anything we can do to help is good. This helps especially with code
-    // that is executed many times quickly, eg tight loops.
-    if (!dataDirToIds.contains(dataDir)) {
-      // Guard against SI-7943: "TrieMap method getOrElseUpdate is not thread-safe".
-      dataDirToIds.synchronized {
-        if (!dataDirToIds.contains(dataDir)) {
-          dataDirToIds(dataDir) = TrieMap.empty[Int, Any]
-        }
-      }
-    }
-    val ids = dataDirToIds(dataDir)
-    if (!ids.contains(id)) {
-      // Each thread writes to a separate measurement file, to reduce contention
-      // and because file appends via FileWriter are not atomic on Windows.
-      var files = threadFiles.get()
-      if (files == null) {
-        files = mutable.HashMap.empty[String, FileWriter]
-        threadFiles.set(files)
+  def invoked(id: Int, dataDir: String): Unit =
+    val set = dataDirToSet.getOrElseUpdate(dataDir, BitSet.empty)
+    if !set.contains(id) then
+      val added = set.synchronized {
```

Comment:
Does this actually improve performance since you have to add a synchronized call?

Reply:
It seems that the synchronized block is small enough and rare enough that it's very rarely under contention.
According to my JMH microbenchmarks, yes, performance is better when only one thread is using the Invoker, but also when 4 threads are using it at the same time.

Reply:
Ah, if you have benchmarks that show an improvement that's great then, consider putting that information in the commit message :).